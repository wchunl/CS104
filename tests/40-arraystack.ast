ROOT "" (0.0.0)
|   TOK_FUNCTION "" (6.22.1)
|   |   TOK_TYPE_ID "" (6.22.1)
|   |   |   TOK_VOID "void" (6.22.1)
|   |   |   TOK_IDENT "fail" (6.22.6)
|   |   TOK_PARAM "(" (6.22.11)
|   |   |   TOK_TYPE_ID "" (6.22.12)
|   |   |   |   TOK_STRING "string" (6.22.12)
|   |   |   |   TOK_IDENT "expr" (6.22.19)
|   |   |   TOK_TYPE_ID "" (6.22.25)
|   |   |   |   TOK_STRING "string" (6.22.25)
|   |   |   |   TOK_IDENT "file" (6.22.32)
|   |   |   TOK_TYPE_ID "" (6.22.38)
|   |   |   |   TOK_INT "int" (6.22.38)
|   |   |   |   TOK_IDENT "line" (6.22.42)
|   TOK_FUNCTION "" (6.24.1)
|   |   TOK_TYPE_ID "" (6.24.1)
|   |   |   TOK_VOID "void" (6.24.1)
|   |   |   TOK_IDENT "putchr" (6.24.6)
|   |   TOK_PARAM "(" (6.24.13)
|   |   |   TOK_TYPE_ID "" (6.24.14)
|   |   |   |   TOK_INT "int" (6.24.14)
|   |   |   |   TOK_IDENT "chr" (6.24.18)
|   TOK_FUNCTION "" (6.25.1)
|   |   TOK_TYPE_ID "" (6.25.1)
|   |   |   TOK_VOID "void" (6.25.1)
|   |   |   TOK_IDENT "putint" (6.25.6)
|   |   TOK_PARAM "(" (6.25.13)
|   |   |   TOK_TYPE_ID "" (6.25.14)
|   |   |   |   TOK_INT "int" (6.25.14)
|   |   |   |   TOK_IDENT "num" (6.25.18)
|   TOK_FUNCTION "" (6.26.1)
|   |   TOK_TYPE_ID "" (6.26.1)
|   |   |   TOK_VOID "void" (6.26.1)
|   |   |   TOK_IDENT "putstr" (6.26.6)
|   |   TOK_PARAM "(" (6.26.13)
|   |   |   TOK_TYPE_ID "" (6.26.14)
|   |   |   |   TOK_STRING "string" (6.26.14)
|   |   |   |   TOK_IDENT "str" (6.26.21)
|   TOK_FUNCTION "" (6.28.1)
|   |   TOK_TYPE_ID "" (6.28.1)
|   |   |   TOK_INT "int" (6.28.1)
|   |   |   TOK_IDENT "getchr" (6.28.5)
|   |   TOK_PARAM "(" (6.28.11)
|   TOK_FUNCTION "" (6.29.1)
|   |   TOK_TYPE_ID "" (6.29.1)
|   |   |   TOK_STRING "string" (6.29.1)
|   |   |   TOK_IDENT "getstr" (6.29.8)
|   |   TOK_PARAM "(" (6.29.14)
|   TOK_FUNCTION "" (6.30.1)
|   |   TOK_TYPE_ID "" (6.30.1)
|   |   |   TOK_STRING "string" (6.30.1)
|   |   |   TOK_IDENT "getln" (6.30.8)
|   |   TOK_PARAM "(" (6.30.13)
|   TOK_STRUCT "struct" (7.7.1)
|   |   TOK_IDENT "stack" (7.7.8)
|   |   TOK_ARRAY "array" (7.8.4)
|   |   |   TOK_TYPE_ID "string" (7.8.10)
|   |   |   TOK_TYPE_ID "data" (7.8.18)
|   |   TOK_INT "int" (7.9.4)
|   |   |   TOK_TYPE_ID "size" (7.9.8)
|   |   TOK_INT "int" (7.10.4)
|   |   |   TOK_TYPE_ID "top" (7.10.8)
|   TOK_FUNCTION "" (7.13.1)
|   |   TOK_TYPE_ID "" (7.13.1)
|   |   |   TOK_PTR "ptr" (7.13.1)
|   |   |   |   TOK_TYPE_ID "stack" (7.13.12)
|   |   |   TOK_IDENT "new_stack" (7.13.19)
|   |   TOK_PARAM "(" (7.13.29)
|   |   |   TOK_TYPE_ID "" (7.13.30)
|   |   |   |   TOK_INT "int" (7.13.30)
|   |   |   |   TOK_IDENT "size" (7.13.34)
|   |   TOK_BLOCK "{" (7.13.40)
|   |   |   TOK_FUNCTION "" (7.14.28)
|   |   |   |   TOK_VARDECL "=" (7.14.28)
|   |   |   |   |   TOK_PTR "ptr" (7.14.4)
|   |   |   |   |   |   TOK_TYPE_ID "stack" (7.14.15)
|   |   |   |   |   |   TOK_TYPE_ID "stack" (7.14.22)
|   |   |   |   |   TOK_ALLOC "alloc" (7.14.30)
|   |   |   |   |   |   TOK_TYPE_ID "stack" (7.14.43)
|   |   |   |   '=' "=" (7.15.16)
|   |   |   |   |   TOK_ARROW "->" (7.15.9)
|   |   |   |   |   |   TOK_IDENT "stack" (7.15.4)
|   |   |   |   |   |   TOK_IDENT "data" (7.15.11)
|   |   |   |   |   TOK_ALLOC "alloc" (7.15.18)
|   |   |   |   |   |   TOK_ARRAY "array" (7.15.24)
|   |   |   |   |   |   TOK_TYPE_ID "string" (7.15.30)
|   |   |   |   |   |   TOK_IDENT "size" (7.15.40)
|   |   |   |   '=' "=" (7.16.16)
|   |   |   |   |   TOK_ARROW "->" (7.16.9)
|   |   |   |   |   |   TOK_IDENT "stack" (7.16.4)
|   |   |   |   |   |   TOK_IDENT "size" (7.16.11)
|   |   |   |   |   TOK_IDENT "size" (7.16.18)
|   |   |   |   '=' "=" (7.17.15)
|   |   |   |   |   TOK_ARROW "->" (7.17.9)
|   |   |   |   |   |   TOK_IDENT "stack" (7.17.4)
|   |   |   |   |   |   TOK_IDENT "top" (7.17.11)
|   |   |   |   |   '-' "-" (7.17.18)
|   |   |   |   |   |   TOK_INTCON "1" (7.17.19)
|   |   |   |   TOK_RETURN "return" (7.18.4)
|   |   |   |   |   TOK_IDENT "stack" (7.18.11)
|   TOK_FUNCTION "" (7.21.1)
|   |   TOK_TYPE_ID "" (7.21.1)
|   |   |   TOK_VOID "void" (7.21.1)
|   |   |   TOK_IDENT "push" (7.21.6)
|   |   TOK_PARAM "(" (7.21.11)
|   |   |   TOK_TYPE_ID "" (7.21.12)
|   |   |   |   TOK_PTR "ptr" (7.21.12)
|   |   |   |   |   TOK_TYPE_ID "stack" (7.21.23)
|   |   |   |   TOK_IDENT "stack" (7.21.30)
|   |   |   TOK_TYPE_ID "" (7.21.37)
|   |   |   |   TOK_STRING "string" (7.21.37)
|   |   |   |   TOK_IDENT "str" (7.21.44)
|   |   TOK_BLOCK "{" (7.21.49)
|   |   |   TOK_FUNCTION "" (7.22.4)
|   |   |   |   TOK_BLOCK "{" (7.22.4)
|   |   |   |   |   TOK_FUNCTION "" (7.22.5)
|   |   |   |   |   |   TOK_IF "if" (7.22.5)
|   |   |   |   |   |   |   TOK_NOT "not" (7.22.9)
|   |   |   |   |   |   |   |   '<' "<" (7.22.25)
|   |   |   |   |   |   |   |   |   TOK_ARROW "->" (7.22.19)
|   |   |   |   |   |   |   |   |   |   TOK_IDENT "stack" (7.22.14)
|   |   |   |   |   |   |   |   |   |   TOK_IDENT "top" (7.22.21)
|   |   |   |   |   |   |   |   |   '-' "-" (7.22.39)
|   |   |   |   |   |   |   |   |   |   TOK_ARROW "->" (7.22.32)
|   |   |   |   |   |   |   |   |   |   |   TOK_IDENT "stack" (7.22.27)
|   |   |   |   |   |   |   |   |   |   |   TOK_IDENT "size" (7.22.34)
|   |   |   |   |   |   |   |   |   |   TOK_INTCON "1" (7.22.41)
|   |   |   |   |   |   |   TOK_CALL "(" (7.22.50)
|   |   |   |   |   |   |   |   TOK_IDENT "fail" (7.22.45)
|   |   |   |   |   |   |   |   TOK_STRINGCON ""stack->top < stack->size - 1"" (7.22.51)
|   |   |   |   |   |   |   |   |   TOK_STRINGCON ""40-arraystack.oc"" (7.22.83)
|   |   |   |   |   |   |   |   |   TOK_INTCON "22" (7.22.103)
|   |   |   |   ';' ";" (7.22.108)
|   |   |   |   '=' "=" (7.23.15)
|   |   |   |   |   TOK_ARROW "->" (7.23.9)
|   |   |   |   |   |   TOK_IDENT "stack" (7.23.4)
|   |   |   |   |   |   TOK_IDENT "top" (7.23.11)
|   |   |   |   |   '+' "+" (7.23.28)
|   |   |   |   |   |   TOK_ARROW "->" (7.23.22)
|   |   |   |   |   |   |   TOK_IDENT "stack" (7.23.17)
|   |   |   |   |   |   |   TOK_IDENT "top" (7.23.24)
|   |   |   |   |   |   TOK_INTCON "1" (7.23.30)
|   |   |   |   '=' "=" (7.24.28)
|   |   |   |   |   TOK_INDEX "[" (7.24.15)
|   |   |   |   |   |   TOK_ARROW "->" (7.24.9)
|   |   |   |   |   |   |   TOK_IDENT "stack" (7.24.4)
|   |   |   |   |   |   |   TOK_IDENT "data" (7.24.11)
|   |   |   |   |   |   TOK_ARROW "->" (7.24.21)
|   |   |   |   |   |   |   TOK_IDENT "stack" (7.24.16)
|   |   |   |   |   |   |   TOK_IDENT "top" (7.24.23)
|   |   |   |   |   TOK_IDENT "str" (7.24.30)
|   TOK_FUNCTION "" (7.27.1)
|   |   TOK_TYPE_ID "" (7.27.1)
|   |   |   TOK_STRING "string" (7.27.1)
|   |   |   TOK_IDENT "pop" (7.27.8)
|   |   TOK_PARAM "(" (7.27.12)
|   |   |   TOK_TYPE_ID "" (7.27.13)
|   |   |   |   TOK_PTR "ptr" (7.27.13)
|   |   |   |   |   TOK_TYPE_ID "stack" (7.27.24)
|   |   |   |   TOK_IDENT "stack" (7.27.31)
|   |   TOK_BLOCK "{" (7.27.38)
|   |   |   TOK_FUNCTION "" (7.28.15)
|   |   |   |   TOK_VARDECL "=" (7.28.15)
|   |   |   |   |   TOK_STRING "string" (7.28.4)
|   |   |   |   |   |   TOK_TYPE_ID "tmp" (7.28.11)
|   |   |   |   |   TOK_INDEX "[" (7.28.28)
|   |   |   |   |   |   TOK_ARROW "->" (7.28.22)
|   |   |   |   |   |   |   TOK_IDENT "stack" (7.28.17)
|   |   |   |   |   |   |   TOK_IDENT "data" (7.28.24)
|   |   |   |   |   |   TOK_ARROW "->" (7.28.34)
|   |   |   |   |   |   |   TOK_IDENT "stack" (7.28.29)
|   |   |   |   |   |   |   TOK_IDENT "top" (7.28.36)
|   |   |   |   TOK_BLOCK "{" (7.29.4)
|   |   |   |   |   TOK_FUNCTION "" (7.29.5)
|   |   |   |   |   |   TOK_IF "if" (7.29.5)
|   |   |   |   |   |   |   TOK_NOT "not" (7.29.9)
|   |   |   |   |   |   |   |   TOK_NE "!=" (7.29.25)
|   |   |   |   |   |   |   |   |   TOK_ARROW "->" (7.29.19)
|   |   |   |   |   |   |   |   |   |   TOK_IDENT "stack" (7.29.14)
|   |   |   |   |   |   |   |   |   |   TOK_IDENT "top" (7.29.21)
|   |   |   |   |   |   |   |   |   '-' "-" (7.29.29)
|   |   |   |   |   |   |   |   |   |   TOK_INTCON "1" (7.29.30)
|   |   |   |   |   |   |   TOK_CALL "(" (7.29.40)
|   |   |   |   |   |   |   |   TOK_IDENT "fail" (7.29.35)
|   |   |   |   |   |   |   |   TOK_STRINGCON ""stack->top != EMPTY"" (7.29.41)
|   |   |   |   |   |   |   |   |   TOK_STRINGCON ""40-arraystack.oc"" (7.29.64)
|   |   |   |   |   |   |   |   |   TOK_INTCON "29" (7.29.84)
|   |   |   |   ';' ";" (7.29.89)
|   |   |   |   '=' "=" (7.30.15)
|   |   |   |   |   TOK_ARROW "->" (7.30.9)
|   |   |   |   |   |   TOK_IDENT "stack" (7.30.4)
|   |   |   |   |   |   TOK_IDENT "top" (7.30.11)
|   |   |   |   |   '-' "-" (7.30.28)
|   |   |   |   |   |   TOK_ARROW "->" (7.30.22)
|   |   |   |   |   |   |   TOK_IDENT "stack" (7.30.17)
|   |   |   |   |   |   |   TOK_IDENT "top" (7.30.24)
|   |   |   |   |   |   TOK_INTCON "1" (7.30.30)
|   |   |   |   TOK_RETURN "return" (7.31.4)
|   |   |   |   |   TOK_IDENT "tmp" (7.31.11)
|   TOK_FUNCTION "" (7.34.1)
|   |   TOK_TYPE_ID "" (7.34.1)
|   |   |   TOK_INT "int" (7.34.1)
|   |   |   TOK_IDENT "empty" (7.34.5)
|   |   TOK_PARAM "(" (7.34.11)
|   |   |   TOK_TYPE_ID "" (7.34.12)
|   |   |   |   TOK_PTR "ptr" (7.34.12)
|   |   |   |   |   TOK_TYPE_ID "stack" (7.34.23)
|   |   |   |   TOK_IDENT "stack" (7.34.30)
|   |   TOK_BLOCK "{" (7.34.37)
|   |   |   TOK_FUNCTION "" (7.35.4)
|   |   |   |   TOK_RETURN "return" (7.35.4)
|   |   |   |   |   TOK_EQ "==" (7.35.22)
|   |   |   |   |   |   TOK_ARROW "->" (7.35.16)
|   |   |   |   |   |   |   TOK_IDENT "stack" (7.35.11)
|   |   |   |   |   |   |   TOK_IDENT "top" (7.35.18)
|   |   |   |   |   |   '-' "-" (7.35.26)
|   |   |   |   |   |   |   TOK_INTCON "1" (7.35.27)
|   TOK_FUNCTION "" (7.38.1)
|   |   TOK_TYPE_ID "" (7.38.1)
|   |   |   TOK_INT "int" (7.38.1)
|   |   |   TOK_IDENT "main" (7.38.5)
|   |   TOK_PARAM "(" (7.38.10)
|   |   |   TOK_TYPE_ID "" (7.38.11)
|   |   |   |   TOK_INT "int" (7.38.11)
|   |   |   |   TOK_IDENT "argc" (7.38.15)
|   |   |   TOK_TYPE_ID "" (7.38.21)
|   |   |   |   TOK_ARRAY "array" (7.38.21)
|   |   |   |   |   TOK_TYPE_ID "string" (7.38.27)
|   |   |   |   TOK_IDENT "argv" (7.38.35)
|   |   TOK_BLOCK "{" (7.38.41)
|   |   |   TOK_FUNCTION "" (7.39.28)
|   |   |   |   TOK_VARDECL "=" (7.39.28)
|   |   |   |   |   TOK_PTR "ptr" (7.39.4)
|   |   |   |   |   |   TOK_TYPE_ID "stack" (7.39.15)
|   |   |   |   |   |   TOK_TYPE_ID "stack" (7.39.22)
|   |   |   |   |   TOK_CALL "(" (7.39.40)
|   |   |   |   |   |   TOK_IDENT "new_stack" (7.39.30)
|   |   |   |   |   |   TOK_INTCON "100" (7.39.41)
|   |   |   |   TOK_VARDECL "=" (7.40.13)
|   |   |   |   |   TOK_INT "int" (7.40.4)
|   |   |   |   |   |   TOK_TYPE_ID "argi" (7.40.8)
|   |   |   |   |   TOK_INTCON "0" (7.40.15)
|   |   |   |   TOK_WHILE "while" (7.41.4)
|   |   |   |   |   '<' "<" (7.41.16)
|   |   |   |   |   |   TOK_IDENT "argi" (7.41.11)
|   |   |   |   |   |   TOK_IDENT "argc" (7.41.18)
|   |   |   |   |   TOK_BLOCK "{" (7.41.24)
|   |   |   |   |   |   TOK_FUNCTION "" (7.42.12)
|   |   |   |   |   |   |   TOK_CALL "(" (7.42.12)
|   |   |   |   |   |   |   |   TOK_IDENT "push" (7.42.7)
|   |   |   |   |   |   |   |   TOK_IDENT "stack" (7.42.13)
|   |   |   |   |   |   |   |   |   TOK_INDEX "[" (7.42.24)
|   |   |   |   |   |   |   |   |   |   TOK_IDENT "argv" (7.42.20)
|   |   |   |   |   |   |   |   |   |   TOK_IDENT "argi" (7.42.25)
|   |   |   |   |   |   |   '=' "=" (7.43.12)
|   |   |   |   |   |   |   |   TOK_IDENT "argi" (7.43.7)
|   |   |   |   |   |   |   |   '+' "+" (7.43.19)
|   |   |   |   |   |   |   |   |   TOK_IDENT "argi" (7.43.14)
|   |   |   |   |   |   |   |   |   TOK_INTCON "1" (7.43.21)
|   |   |   |   TOK_WHILE "while" (7.45.4)
|   |   |   |   |   TOK_NOT "not" (7.45.11)
|   |   |   |   |   |   TOK_CALL "(" (7.45.21)
|   |   |   |   |   |   |   TOK_IDENT "empty" (7.45.15)
|   |   |   |   |   |   |   TOK_IDENT "stack" (7.45.22)
|   |   |   |   |   TOK_BLOCK "{" (7.45.30)
|   |   |   |   |   |   TOK_FUNCTION "" (7.46.14)
|   |   |   |   |   |   |   TOK_CALL "(" (7.46.14)
|   |   |   |   |   |   |   |   TOK_IDENT "putstr" (7.46.7)
|   |   |   |   |   |   |   |   TOK_CALL "(" (7.46.19)
|   |   |   |   |   |   |   |   |   TOK_IDENT "pop" (7.46.15)
|   |   |   |   |   |   |   |   |   TOK_IDENT "stack" (7.46.20)
|   |   |   |   |   |   |   TOK_CALL "(" (7.47.14)
|   |   |   |   |   |   |   |   TOK_IDENT "putchr" (7.47.7)
|   |   |   |   |   |   |   |   TOK_CHARCON "'\n'" (7.47.15)
|   |   |   |   TOK_RETURN "return" (7.49.4)
|   |   |   |   |   TOK_INTCON "0" (7.49.11)
