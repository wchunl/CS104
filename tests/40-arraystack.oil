fail:     .function
          .param ptr expr
          .param ptr file
          .param int line
          .end
putchr:   .function
          .param int chr
          .end
putint:   .function
          .param int num
          .end
putstr:   .function
          .param ptr str
          .end
getchr:   .function int
          .end
getstr:   .function
          .end
getln:    .function
          .end
          .struct stack
          .field ptr string
          .field int size
          .field int top
          .end
new_stack:.function
          .param int size
          .local ptr stack
          malloc (size of struct stack)
          stack = $t0:i
          malloc(4 * sizeof ptr)
          stack->stack.data = $t1:i
          stack->stack.size = size
          $t2:i = - 1
          stack->stack.top = $t2:i
          return stack
          return 
          .end
push:     .function
          .param ptr stack
          .param ptr str
.if0:     $t3:p = stack->stack.size
          $t4:i = $t3:i - 1
          $t5:i = $t4:i < -
          $t6:i = not $t5:i
          goto .el0 if not $t6:i
.th0:     $t7:i = "stack->top < stack->size - 1"
          fail($t7:i)
.fi0:     $t8:p = stack->stack.top
          $t9:i = $t8:i + 1
          stack->stack.top = $t9:i
          ->[-> * :p] = str
          .end
pop:      .function
          .param ptr stack
          .local ptr tmp
          $t10:p = ->[-> * :p]
          tmp = $t10:i
.if1:     $t11:i = - 1
          $t12:i = $t11:i != -
          $t13:i = not $t12:i
          goto .el1 if not $t13:i
.th1:     $t14:i = "stack->top != EMPTY"
          fail($t14:i)
.fi1:     $t15:p = stack->stack.top
          $t16:i = $t15:i - 1
          stack->stack.top = $t16:i
          return tmp
          return 
          .end
empty:    .function int
          .param ptr stack
          return ==
          return 
          .end
main:     .function int
          .param int argc
          .param ptr argv
          .local ptr stack
          .local int argi
          new_stack(100)
          stack = $t17:i
          argi = 0
.wh0:     $t18:i = argi < argc
          goto .od0 if not $t18:i
.do0:     $t19:i = stack
          push($t19:i)
          $t20:i = argi + 1
          argi = $t20:i
.od0:     
.wh1:     empty(stack)
          $t22:i = not $t21:i
          goto .od1 if not $t22:i
.do1:     pop(stack)
          putstr($t23:i)
          putchr('\n')
.od1:     return 0
          return 
          .end
