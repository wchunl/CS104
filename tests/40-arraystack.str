string_set[   2]:    5320480286546035437 0x10ab448->"stack"
                   12597418120160561823 0x10a9b68->"int"
string_set[   5]:    1370730858159036685 0x10ad9f8->"+"
string_set[   8]:   13497578857207247412 0x10a9a28->"string"
string_set[   9]:   10319632053171904911 0x10b07e8->"while"
                   13653664462152793141 0x10af298->"29"
string_set[  10]:    9201800968376702500 0x10aac58->"getchr"
string_set[  11]:   14805121035671447500 0x10ac2f8->"->"
string_set[  12]:   14643508047410943861 0x10b0438->"argv"
                   14752253762927134915 0x10ab848->"data"
string_set[  14]:   15816506932648033945 0x10a5228->"("
                    1396094175042499165 0x10ab6e8->"<"
string_set[  15]:   12868445110721718657 0x10a9d38->")"
string_set[  17]:    6142509188972423790 0x10a5588->""
                    5111042995199451563 0x10ad748->""stack->top < stack->size - 1""
string_set[  20]:    1583576014053361350 0x10aa8a8->"str"
string_set[  21]:    9317282556992397945 0x10aced8->"if"
string_set[  22]:   10159970873491820195 0x10ac978->"1"
string_set[  23]:     232592375681467293 0x10ae7d8->"tmp"
string_set[  24]:     370180285156928338 0x10af838->"empty"
                    8939588259793473717 0x10ad848->""40-arraystack.oc""
string_set[  26]:   12291392672316406109 0x10a9ec8->"line"
                    9611428142580219387 0x10a55a8->"fail"
                   11597697714117577063 0x10abff8->"="
string_set[  28]:   18413295556186398326 0x10aa118->"putint"
string_set[  29]:     549611050806166819 0x10ae3a8->"pop"
                   11911796782241914524 0x10a9688->"void"
string_set[  33]:    6814028903552918164 0x10aa078->"array"
string_set[  34]:   12099763674597357221 0x10af128->""stack->top != EMPTY""
string_set[  35]:   11868120863286193964 0x10affb8->"main"
string_set[  40]:   13945322250941355849 0x10ab2d8->"struct"
string_set[  41]:   12051031859636400188 0x10ab588->"]"
string_set[  42]:   13686721134187162920 0x10aba98->"}"
string_set[  43]:   14849783028996024063 0x10ad208->"not"
string_set[  49]:   12679825168387489523 0x10aa728->"num"
string_set[  50]:      21606623943142546 0x10aa058->"{"
string_set[  52]:    5344662657899890615 0x10a9ee8->";"
string_set[  54]:    7296690267450279874 0x10aa388->"chr"
                    2497302301362042119 0x10abc98->"new_stack"
string_set[  55]:    2297668033614959926 0x10b09f8->"0"
                   17176659130743002837 0x10abb38->"ptr"
string_set[  56]:    3701972582333163920 0x10aff78->"argc"
string_set[  59]:    6192799629375248351 0x10b1728->"'\n'"
string_set[  60]:    1815486778388015849 0x10acbc8->"push"
string_set[  63]:   14526079891305879776 0x10ab7d8->">"
string_set[  64]:   16579916199511856337 0x10ab0b8->"getln"
                   13934964484049229816 0x10ab7f8->"size"
string_set[  67]:    4706206694200543916 0x10a9c28->","
string_set[  69]:    5266873771696973564 0x10b08e8->"argi"
string_set[  73]:    9386170080689500517 0x10a9af8->"expr"
string_set[  75]:   16247530808313002034 0x10ab8b8->"top"
string_set[  76]:   11166701219533395393 0x10aae78->"getstr"
string_set[  79]:    6340238197210063670 0x10b0218->"100"
string_set[  80]:    2617056156435215158 0x10aee38->"!="
                    8290819294854170237 0x10ac068->"alloc"
                    7211130878751977441 0x10aa248->"putchr"
string_set[  83]:   13581035385482130682 0x10ada18->"["
                    8832308542017959627 0x10ad8d8->"22"
string_set[  85]:    4017852018407011111 0x10a9d18->"file"
                    7372855250431704371 0x10ac6a8->"-"
string_set[  87]:   14635384904416252381 0x10aca38->"return"
string_set[  89]:   13348923631108115385 0x10aa508->"putstr"
string_set[  95]:   14736352883882276958 0x10afcd8->"=="
load_factor = 0.649
bucket_count = 97
max_bucket_size = 3
