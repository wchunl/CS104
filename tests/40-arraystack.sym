fail (6.22.1) {0} void function
   expr (6.22.12) {1} string variable param lval 0
   file (6.22.25) {1} string variable param lval 1
   line (6.22.38) {1} int variable param lval 2

putchr (6.24.1) {1} void function
   chr (6.24.14) {1} int variable param lval 0

putint (6.25.1) {2} void function
   num (6.25.14) {1} int variable param lval 0

putstr (6.26.1) {3} void function
   str (6.26.14) {1} string variable param lval 0

getchr (6.28.1) {4} int function

getstr (6.29.1) {5} string function

getln (6.30.1) {6} string function

stack (7.7.1) {0} struct stack
   string (7.8.4) array <string> field 0
   size (7.9.4)  int field 1
   top (7.10.4)  int field 2

new_stack (7.13.1) {7} ptr <struct stack> function
   size (7.13.30) {1} int variable param lval 0
   stack (7.14.28) {1} ptr <struct stack> variable local lval 0

push (7.21.1) {8} void function
   stack (7.21.12) {1} ptr <struct stack> variable param lval 0
   str (7.21.37) {1} string variable param lval 1

pop (7.27.1) {9} string function
   stack (7.27.13) {1} ptr <struct stack> variable param lval 0
   tmp (7.28.15) {1} string variable local lval 0

empty (7.34.1) {10} int function
   stack (7.34.12) {1} ptr <struct stack> variable param lval 0

main (7.38.1) {11} int function
   argc (7.38.11) {1} int variable param lval 0
   argv (7.38.21) {1} array <string> variable param lval 1
   stack (7.39.28) {1} ptr <struct stack> variable local lval 0
   argi (7.40.13) {1} int variable local lval 1
