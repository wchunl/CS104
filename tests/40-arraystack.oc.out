; 0.  1: # 1 "40-arraystack.oc"; 1.  0: 
; 1.  1: # 1 "<built-in>"
; 2.  1: # 1 "<command-line>"
; 3.  1: # 1 "40-arraystack.oc"
; 4.  1: 
; 4.  2: 
; 4.  3: # 1 "oclib.h" 1
; 5.  1: # 22 "oclib.h"
; 6. 22: void fail (string expr, string file, int line);
; 6. 23: 
; 6. 24: void putchr (int chr);
; 6. 25: void putint (int num);
; 6. 26: void putstr (string str);
; 6. 27: 
; 6. 28: int getchr();
; 6. 29: string getstr();
; 6. 30: string getln();
; 6. 31: # 4 "40-arraystack.oc" 2
; 7.  4: 
; 7.  5: 
; 7.  6: 
; 7.  7: struct stack {
; 7.  8:    array<string> data;
; 7.  9:    int size;
; 7. 10:    int top;
; 7. 11: };
; 7. 12: 
; 7. 13: ptr<struct stack> new_stack (int size) {
; 7. 14:    ptr<struct stack> stack = alloc<struct stack>();
; 7. 15:    stack->data = alloc<array<string>> (size);
; 7. 16:    stack->size = size;
; 7. 17:    stack->top = (-1);
; 7. 18:    return stack;
; 7. 19: }
; 7. 20: 
; 7. 21: void push (ptr<struct stack> stack, string str) {
; 7. 22:    {if (not (stack->top < stack->size - 1)) fail ("stack->top < stack->size - 1", "40-arraystack.oc", 22);};
; 7. 23:    stack->top = stack->top + 1;
; 7. 24:    stack->data[stack->top] = str;
; 7. 25: }
; 7. 26: 
; 7. 27: string pop (ptr<struct stack> stack) {
; 7. 28:    string tmp = stack->data[stack->top];
; 7. 29:    {if (not (stack->top != (-1))) fail ("stack->top != EMPTY", "40-arraystack.oc", 29);};
; 7. 30:    stack->top = stack->top - 1;
; 7. 31:    return tmp;
; 7. 32: }
; 7. 33: 
; 7. 34: int empty (ptr<struct stack> stack) {
; 7. 35:    return stack->top == (-1);
; 7. 36: }
; 7. 37: 
; 7. 38: int main (int argc, array<string> argv) {
; 7. 39:    ptr<struct stack> stack = new_stack (100);
; 7. 40:    int argi = 0;
; 7. 41:    while (argi < argc) {
; 7. 42:       push (stack, argv[argi]);
; 7. 43:       argi = argi + 1;
; 7. 44:    }
; 7. 45:    while (not empty (stack)) {
; 7. 46:       putstr (pop (stack));
; 7. 47:       putchr ('\n');
; 7. 48:    }
; 7. 49:    return 0;
; 7. 50: }
hello